feat(#1932): implement smooth scrolling for anchor navigation across site

Mohit5Upadhyay opened the pull request:
<!--
Thank you for your pull request. Please provide a description and 
note the Certificate of Origin below. 

-->

## Description
This PR implements smooth scrolling for in-page navigation across the Express.js website as mentioned in Issue : #1932 

This improves the user experience by ensuring that navigation to sections via anchor links transitions smoothly, rather than jumping abruptly.

Closes #1932 


## Video

https://github.com/user-attachments/assets/0efeb664-b772-446d-98d3-b033cf68abb8


<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

Mohit5Upadhyay committed:
feat(#1932): implement smooth scrolling for anchor navigation across site

ShubhamOulkar commented on js/app.js:
```diff
@@ -1,12 +1,56 @@
+  $(document).on('click', 'a[href^="#"]:not([href="#"]):not([href="#top"])', function(e) {
+    var href = $(this).attr('href');
+    if (smoothScrollToElement(href)) {
+      e.preventDefault();
```
Can you please remove this line https://github.com/expressjs/expressjs.com/blob/gh-pages/js/menu.js#L186 ? Since default behavior stopped here.

ShubhamOulkar commented on js/app.js:
```diff
@@ -1,12 +1,56 @@
   });
 
+  // Smooth scrolling for all anchor links
+  $(document).on('click', 'a[href^="#"]:not([href="#"]):not([href="#top"])', function(e) {
```
```suggestion
  $("#api-doc").on('click', 'a[href^="#"]:not([href="#"]):not([href="#top"])', function(e) {
```
We should minimize the scope of this click event. Currently, $(document).on('click'...) attaches the handler to all anchor links with `#` across every page. However, smooth scrolling is only needed on API pages, since the TOC exists only there. Scoping it to `#api-doc` avoids unnecessary event handling and prevents inconsistent anchor `#` behavior on other pages. Also, checking `"#api-doc"` exist on page is required before applying click event.

bjohansebas reviewed (changes requested):
Thanks for the PR!

As i mentioned in the issue, this can be handled through CSS, javascript is not necessary.

ShubhamOulkar commented:
> this can be handled through CSS,

I am OK with CSS only solution. But this approach comes with some limitations, like we cannot control scroll behavior and can not add header offsets. But it is fine if headings are not hiding behind the header bar on anchor link click.

Mohit5Upadhyay commented:
Thanks @bjohansebas for the suggestion about a CSS-only solution!
Unfortunately, I think for this project, CSS alone isn’t sufficient because:
- IDs like #app.METHOD `(http://localhost:4000/en/5x/api.html#app.listen)` contain dots, which CSS can’t reliably scroll to without custom JS.
- We need a precise scroll offset (120px) to avoid headings hiding behind the navbar.

So I’d propose keeping the JS smooth scroll, scoped only to `#api-doc` to avoid global impact as per suggested by @ShubhamOulkar.

Let me know if you’d prefer another approach!
Thanks again for your feedback.

bjohansebas commented:
>But it is fine if headings are not hiding behind the header bar on anchor link click.

The headings are never hidden by the header bar, or at least I’ve been trying to reproduce it and haven’t been able to.

> IDs like #app.METHOD (http://localhost:4000/en/5x/api.html#app.listen) contain dots, which CSS can’t reliably scroll to without custom JS.

That doesn’t matter in CSS, since the effect will be applied to all elements anyway.
```css
* {
  scroll-behavior: smooth;
}
```

Mohit5Upadhyay committed:
feat: implement CSS-only smooth scrolling for anchor navigation

Mohit5Upadhyay commented:
Hi @bjohansebas and @ShubhamOulkar,

I’ve updated the implementation to follow your suggestion of a CSS-only smooth scrolling solution.
- I removed all the custom JavaScript smooth scroll code I previously added.
- Instead, I’ve applied this in CSS:
```
* {
  scroll-behavior: smooth;
}

h1, h2, h3, h4, h5, h6,
[id] {
  scroll-margin-top: 120px;
}

@media (prefers-reduced-motion: reduce) {
  html {
    scroll-behavior: auto;
  }
}
```

Let me know if you’d like any further changes!
Thanks for your time! 🙏

bjohansebas commented on css/style.css:
```diff
@@ -1,6 +1,20 @@
+}
+
+/* handle header offset with scroll-margin */
+h1, h2, h3, h4, h5, h6,
+[id] {
+  scroll-margin-top: 120px;
+}
```
I'm not sure if this is actually doing anything, and if it's supposed to have the same effect as in the MDN example https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-margin-top, then I'm -1 on this, I don't think we should do it.

ShubhamOulkar replied:
| without scroll top margin | with scroll top margin |
|------|-------|
|<img width="1242" height="372" alt="image" src="https://github.com/user-attachments/assets/0a84047a-768a-49cd-a595-1cb7bedbbe7c" />|<img width="1297" height="385" alt="image" src="https://github.com/user-attachments/assets/c8aea1ab-0c19-43fa-a859-4a28514ca04c" />|

We can clearly see that margin spaces added to the header on scroll from the top of the viewport.
This property is used to increase visible area around the target element so that it is accessible to the user. 

In our context, why it is important? 
I expressed my concerns over the issue of hiding headers behind the top navigation bar in a comment above. 
> But it is fine if headings are not hiding behind the header bar on anchor link click.

This happens if top navigation is positioned sticky at the top. See image, `express.raw()` hide behind sticky top navigation.
<img width="1920" height="469" alt="image" src="https://github.com/user-attachments/assets/4f47bf69-4701-403c-a9ef-6cb007673f43" />

It is not required right now, without scroll top margin, header scroll into viewport at top 0px. But I think adding some small margin 60px to the top looks good for me. Also our header is 57px height. Maybe I will go with sticky top navigation bar in #1943. 
<img width="1625" height="422" alt="image" src="https://github.com/user-attachments/assets/17f56f22-0703-413c-aeef-24885dfcd6e2" />

[see scroll_padding_and_margin](https://web.dev/articles/css-scroll-snap#scroll_padding_and_margin)

ShubhamOulkar commented on css/style.css:
```diff
@@ -1,6 +1,20 @@
+/* handle header offset with scroll-margin */
+h1, h2, h3, h4, h5, h6,
+[id] {
+  scroll-margin-top: 120px;
```
```suggestion
  scroll-margin-top: 60px;
```
see comment https://github.com/expressjs/expressjs.com/pull/1965/files#r2206464995

ShubhamOulkar commented on css/style.css:
```diff
@@ -264,11 +278,6 @@ a {
   padding-top: 40px;
 }
 
-#api-doc *:target, #page-doc *:target {
```
I support removing this
since we are using scroll top margin only on headers.

ShubhamOulkar commented on css/style.css:
```diff
@@ -1,6 +1,20 @@
+}
+
+/* Respect user motion preferences for access */
+@media (prefers-reduced-motion: reduce) {
```
It is Good practice to consider accessibility
+1

ShubhamOulkar reviewed (changes requested):
Please do remove this line https://github.com/expressjs/expressjs.com/blob/gh-pages/js/menu.js#L186. On TOC main section heading click, it should scroll to main sections on the page. Currently, it is disabled. Only subsections get scrolled.

Mohit5Upadhyay committed:
apply reviewer changes: Update scroll-margin-top to 60px and remove preventDefault for TOC links

Mohit5Upadhyay commented:
Hi @bjohansebas and @ShubhamOulkar,

Thanks so much for the thoughtful reviews!

Regarding `scroll-margin-top`:
I agree with @ShubhamOulkar that it’s helpful to prevent headings from hiding behind the sticky nav bar.
> But I think adding some small margin 60px to the top looks good for me. Also our header is 57px height. Maybe I will go with sticky top navigation bar in https://github.com/expressjs/expressjs.com/issues/1943.

I’ll adjust the value to 60px as suggested, since our header height is ~57px.

Remove `event.preventDefault()` in menu.js so main TOC sections scroll as expected.

Let me know if there’s anything else you’d like tweaked!
Thanks again for your guidance.

ShubhamOulkar reviewed (commented):
I just tested, but default CSS scroll is too fast. On clicking main menu, the page scrolls really fast to that menu. Should we try putting some animation key frames to add some delays to scrolls?

Mohit5Upadhyay commented:
hi @ShubhamOulkar,

Thank you for the suggestion regarding adding delays to scroll using animation keyframes.

I explored the possibility of controlling scroll speed using CSS keyframes alone. However, after testing locally, it turns out that CSS keyframes cannot manipulate scroll behavior that we need. 

While `scroll-behavior: smooth` provides a basic smooth scroll effect, it doesn’t allow customization of speed or delay — and keyframes don’t apply to scroll position singly to achieve this. 

To achieve the effect you’re suggesting (slower or delayed scrolling), we would need to implement a small JavaScript solution that programmatically controls the scroll animation.

Please let me know if you'd like me to proceed with a JavaScript-based implementation to match the desired behavior, or if you have any other suggestions — I’d be grateful to explore them!

Thanks!

bjohansebas commented:
> Please let me know if you'd like me to proceed with a JavaScript-based implementation to match the desired behavior, or if you have any other suggestions — I’d be grateful to explore them!

I’m -1 on implementing this in JavaScript. I want to simplify things as much as possible, and I don’t see it as reasonable to maintain that JavaScript when the platform itself already provides this solution (it might not be perfect, but it’s better than having to maintain JavaScript).

bjohansebas commented on js/menu.js:
```diff
@@ -183,7 +183,6 @@ toggleBtn?.addEventListener("click", (e) => {
 // Open/Close sub TOC content on click
 document.querySelectorAll("#menu > li > a").forEach((link) => {
 	link.addEventListener("click", function (event) {
-	  event.preventDefault(); // stop navigation to submenu
```
why?

bjohansebas replied:
This needs to stay. I've been testing, and on mobile it doesn’t work correctly, with this line, it works.

ShubhamOulkar replied:
This line stop scrolling to submenu section headings. E.g. Request,  Response,  etc. 
Actually here two scrolling actions happening inside TOC and on the page.

ShubhamOulkar replied:
Yeh on mobile it is very bad experience

ShubhamOulkar replied:
We can stop this on small screens by using line 4 menu.js. Let's focus on desktop. It is important for desktop experience. 
 @Mohit5Upadhyay could you try to control behavior in side menu bar? Slow down submenu opening ans scrolling. I think we can do this in CSS. 😃

bjohansebas committed:
Merge branch 'gh-pages' into feat-1932-smooth-scrolling

bjohansebas committed:
remove scroll to top functionality and prevent default navigation for submenu links

Signed-off-by: Sebastian Beltran <bjohansebas@gmail.com>

bjohansebas reviewed (approved):
LGTM

ShubhamOulkar reviewed (approved):
Need to remove line no 5 from app.js. Otherwise LGTM 👍
